osquery/tables/system/linux/pci_devices.cpp

```diff
@@ -237,6 +238,143 @@ Status PciDB::getSubsystemInfo(const std::string& vendor_id,
+  row.emplace("subsystem_vendor_id", "0x" + subsys_vendor_id);
+  row.emplace("subsystem_model_id", "0x" + subsys_model_id);
+
+  std::string content;
+  if (pcidb.getVendorName(subsys_vendor_id, content).ok()) {
+    row.emplace("subsystem_vendor", std::move(content));
+  }
+
+  content.clear();
+  if (pcidb
+          .getSubsystemInfo(
+              vendor_id, model_id, subsys_vendor_id, subsys_model_id, content)
+          .ok()) {
+    row.emplace("subsystem_model", std::move(content));
+  }
+}
+
+Status extractVendorModelFromPciDBIfPresent(Row& row,
+                                            std::string device_ids_attr,
+                                            std::string subsystem_ids_attr,
+                                            const PciDB& pcidb) {
+  std::string vendor_id;
+  std::string model_id;
+  auto status = splitVendorModelAttrs(device_ids_attr, vendor_id, model_id);
+  if (!status.ok()) {
+    // Legacy behavior of using value "0" being supported for backward
+    // compatability.
+    row.emplace("vendor_id", "0");
+    row.emplace("model_id", "0");
+
+    return Status::failure("Failed to parse PCI device ID attributes: " +
+                           status.getMessage());
+  }
+
+  extractOEMVendorModelFromPciDB(row, vendor_id, model_id, pcidb);
+
+  std::string subsystem_vendor_id;
+  std::string subsystem_model_id;
+  status = splitVendorModelAttrs(
+      subsystem_ids_attr, subsystem_vendor_id, subsystem_model_id);
+  if (!status.ok()) {
+    return Status::failure(
+        "Failed to parse PCI device subsystem ID attributes: " +
+        status.getMessage());
+  }
+
+  extractSubsysVendorModelFromPciDB(
+      row, vendor_id, model_id, subsystem_vendor_id, subsystem_model_id, pcidb);
+
+  return Status::success();
+}
+
+Status extractPCIVendorModelInfo(
+    Row& row,
+    std::unique_ptr<udev_device, decltype(&udev_device_unref)>& device,
+    const PciDB& pcidb) {
+  // Fallback data comes from UdevEventPublisher.
+  row["vendor"] = UdevEventPublisher::getValue(device.get(), kPCIKeyVendor);
+  row["model"] = UdevEventPublisher::getValue(device.get(), kPCIKeyModel);
+
+  // Now try PciDB for more up to date info.
+  return extractVendorModelFromPciDBIfPresent(
+      row,
+      UdevEventPublisher::getValue(device.get(), kPCIKeyID),
+      UdevEventPublisher::getValue(device.get(), kPCISubsysID),
+      pcidb);
+}
+
+Status extractPCIClassIDAttrs(Row& row, std::string pci_class_attr) {
+  // pci.ids lower cases everything, so we follow suit.
+  boost::algorithm::to_lower(pci_class_attr);
+
+  auto id_len = pci_class_attr.length();
+  switch (id_len) {
+  case 5:
+    row.emplace("pci_class_id", "0x0" + pci_class_attr.substr(0, 1));
+    row.emplace("pci_subclass_id", "0x" + pci_class_attr.substr(1, 2));
+    break;
+
+  case 6:
+    row.emplace("pci_class_id", "0x" + pci_class_attr.substr(0, 2));
+    row.emplace("pci_subclass_id", "0x" + pci_class_attr.substr(2, 2));
```

Comment:
it looks like a copy paste from the code above, can we avoid that with some intermediate variables for `pci_class_id`, `pci_subclass_id` ?

Reply:
Are you talking about line `360`?